docs: unit test doc (#422)
move-coop/parsons: docs/write_tests.rst

> How to Write Tests for Parsons Connectors

Reviewer: Is this doc linked from How to Build a Connector?  We should probably stick a quick link in the 'adding tests' section.

=========================================

******************
How to Write Tests
******************

Most of our Parsons Connector classes fall into two categories:

 * Classes that call directly into HTTP API's using the Parsons ``APIConnector`` class
 * Classes that wrap around third party Python libraries, often using their own "client" class

> The category that the class falls into generally determines how one will write unit tests for the class.  However, the category of Connector determines how we simulate the external API.

Reviewer: I'm having trouble parsing this, specifically that second sentence. What are you trying to say here?

^^^^^^^^^^^^^^^
Getting Started
^^^^^^^^^^^^^^^

 * Add a folder *test_yourconnectorname* in parsons/test for your connector
 * Add a file *test_yourconnectorname.py* to the *test_yourconnectorname* folder
 * Add one `“Happy Path” <https://en.wikipedia.org/wiki/Happy_path>`_ test per public method of your connector

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Tests for HTTP API Connectors
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For classes built using the ``APIConnector`` class, we can leverage the ``requests-mock`` library, since the ``APIConnector`` uses the ``requests`` library.

The ``requests-mock`` library allows us to simulate the API we are calling into by pretending to be an HTTP server. We can create canned responses, and introspect on the endpoints that our Connector called to ensure that the method we are testing is acting the way we expect.

For an example of testing a class built on the ``APIConnector``, we can look at the ``MailChimp`` connector.

In the code below, we have our ``TestMailchimp`` that serves as our test case. We have one test method to test calls to our ``get_campaigns`` method in our ``Mailchimp`` Connector class.

> We decorate ``test_get_campaigns`` with the `requests_mock.Mocker`` class, which passes in our requests mocker as an argument ``m``. We can then use this mock to configure our mock server.

Reviewer: Missing a tick around requests_mock.Mocker

Below, we will use the ``get`` method to tell the mock HTTP server that we expect a call to the ``campaigns`` endpoint of our API, and that the mock HTTP server should return our ``test_campaignss` canned response. The Mailchimp tests store the canned responses in a separate Python module, which can help keep the test code separate from the test data, which helps makes tests more readable.

.. code-block:: python

    class TestMailchimp(unittest.TestCase):

        @requests_mock.Mocker()
        def test_get_campaigns(self, m):

            mc = MailChimp(api_key='12345')

            # Test that campaigns are returned correctly.
            m.get(self.mc.uri + 'campaigns', json=expected_json.test_campaigns)
            tbl = self.mc.get_campaigns()

            self.assertEqual(tbl.num_rows, 2)

> After wiring up the mock HTTP API, we are ready to call the ``get_campaigns`` method on our ``Mailchimp`` connector. Since we are using the ``requests_mock.Mocker`` class, our Connector will not actually hit the Mailchimp API; our call will be intercepted and passed to our mock HTTP server.

Reviewer 2: It might be a bit confusing to say "our mock HTTP server" as it isn't really a mock http server, but a canned response. Perhaps something like
```suggestion
After wiring up the mock HTTP API, we are ready to call the ``get_campaigns`` method on our ``Mailchimp`` connector. Since we are using the ``requests_mock.Mocker`` class, our Connector will not actually hit the Mailchimp API; our call will be intercepted and the configured canned response will be returned.
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Tests for Connectors Built on Third Party Libraries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For classes that wrap third party libraries, we are looking to simulate whatever client object that serves as the Python API for the calls to the external service. To simulate the client object, we can use the ``unittest.mock.MagicMock`` class from the Python standard library.

The ``MagicMock`` class allows us to build a fake version of the client object by simulating functionality we expect our Connector class to use -- namely by returning canned responses to method calls on the client. As well, the ``MagicMock`` class is recording those calls as they happen, so that after we have run through our tests, we can check that things happened in the way we expected.

> The following gives a simple example of initializing, setting up, and using a ``MagicMock`` class:

Reviewer 2: This might lead to some confusion about "using" the mock, since the example code here is focused on setting up and checking expectations. Calling the mock directly like this is just for illustrative value.

```suggestion
The following gives a simple example of initializing, setting up, and checking expectations on a ``MagicMock`` class. Please see the following, SalesforceTest example on integrating it with your Connector class.
```

.. code-block:: python

    from unittest.mock import MagicMock

    # Initialize a MagicMock object to serve as our fake client
    mock_client = MagicMock()
    # Tell the client to return a canned list of people from its list_people method
    mock_client.list_people.return_value = [
        {'id': 1, 'name': 'Nicole Jackson'},
    ]

>     # Call our mock client's list_people method
>     people = mock_client.list_people()

Reviewer 2: Emphasize that this is where your Connector would call it.
```suggestion
    # Your Connector would call the client like normal
    people = mock_client.list_people()
```

>     # Check that we got our expected data
>     assert len(people) == 1
>     assert people[0]['id'] == 1

Reviewer 2: I would put some spacing here to highlight that this moves into the assertions.
```suggestion

    # Check that we got our expected data
    assert len(people) == 1
    assert people[0]['id'] == 1
```
